File: 백준/Gold/1753. 최단경로/최단경로.py
import sys
INF=sys.maxsize

n,m=map(int,sys.stdin.readline().split())
startVertex=int(sys.stdin.readline())

distance=[INF]*(n+1)
visited=[False]*(n+1)

graph=[[] for _ in range(n+1)]

for _ in range(m):
    start,end,weight=map(int,sys.stdin.readline().split())
    graph[start].append((end,weight))

def get_min_pos():
    min_value=INF
    minpos=-1

    for i in range(1,n+1):
        if visited[i]==0 and distance[i]<=min_value:
            min_value=distance[i]
            minpos=i

    return minpos

def Dijkstra(start):

    for i in graph[start]:
        if distance[i[0]]>=i[1]:
            distance[i[0]]=i[1]

    distance[start]=0
    visited[start]=True

    for _ in range(n-1):
        w=get_min_pos()
        visited[w]=True

        for i in graph[w]:
            if distance[w]+i[1] < distance[i[0]]:
                distance[i[0]]= distance[w]+i[1]

Dijkstra(startVertex)

for i in range(1,n+1):
    if distance[i]==INF:
        print('INF')
    else:
        print(distance[i])

File: 백준/Gold/1753. 최단경로/최단경로.cc
#include <iostream>
#include <vector>
#include <queue>
#include <cstring>

#define MAX_SIZE 987654321
using namespace std;

vector<pair<int, int>> arr[20004];
int dist[20004];
int visited[20004];
int V,E,start;

void init_() {
    for(int i=0;i<20004;i++) {
        dist[i]=MAX_SIZE;
    }
}

void Dijkstra(int start) {
    
    priority_queue<pair<int, int>,vector<pair<int, int>>,greater<pair<int, int>>> q;
    dist[start]=0;
    q.push({0,start});
    
    while(!q.empty()) {
        start=q.top().second;
        q.pop();
        visited[start]=1;
        
        for(pair<int, int> temp:arr[start]) {
            if(visited[temp.second]==0) {
                if(dist[start]+temp.first<dist[temp.second]) {
                    dist[temp.second]=dist[start]+temp.first;
                    q.push({dist[start]+temp.first,temp.second});
                }
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    
    cin >> V >> E;
    cin >> start;
    
    
    for(int i=0;i<E;i++) {
        int start,end,weight;
        cin >> start >> end >> weight;
        arr[start].push_back({weight,end});
    }
    
    init_();
    Dijkstra(start);
    
    for(int i=1;i<=V;i++) {
        
        if(dist[i]==MAX_SIZE) {
            cout << "INF" << "\n";
        }
        else {
            cout << dist[i] << "\n";
        }
    }
    
    return 0;
}
